sql: flake in TestTenantLogic_temp_table/create_after_discard_and_drop_database

Seen on seemingly unrelated bors batch [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=12685076&buildTypeId=Cockroach_BazelEssentialCi):
```
Failed
=== RUN   TestTenantLogic_temp_table/create_after_discard_and_drop_database
    logic.go:2177: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/4937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/temp_table:395: 
        expected success, but found
        (42P01) relation "pg_temp.t" does not exist
        errors.go:167: in NewUndefinedRelationError()
[02:11:25] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/4937/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/temp_table with config 3node-tenant: 89 tests, 1 failures
    --- FAIL: TestTenantLogic_temp_table/create_after_discard_and_drop_database (0.48s)
```


Jira issue: CRDB-33524

I haven't managed to figure out exactly what's causing the problem here. Reverting https://github.com/cockroachdb/cockroach/commit/45b8d05db3ce6f15e4da3cc5b27181d20dbef34a seems to prevent the flake, but all that commit does is avoid planning a constrained scan in some places, which shouldn't affect correctness. The issue still reproduces with the previous commit reverted, so the bug can't have been introduced then. I saw another worrying failure a few times while running the test locally:
```
    --- FAIL: TestTenantLogic_temp_table/regression_108751 (45.27s)
        logic.go:2966: let $t_108751_view_id = 157
        logic.go:2908:

            /private/var/tmp/_bazel_andrewkimball/040b9fbb4b2ac9dbefeac185f50a8f1c/sandbox/darwin-sandbox/5165/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin_arm64-fastbuild/bin/pkg/ccl/logictestccl/tests/3node-tenant/3node-tenant_test_/3node-tenant_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/temp_table:458: SELECT descriptor FROM system.descriptor WHERE id = 157
            expected:

            but found (query options: "retry") :
                [10 179 2 10 13 116 95 49 48 56 55 53 49 95 118 105 101 119 24 157 1 32 154 1 40 1 58 0 66 38 10 1 120 16 1 26 12 8 1 16 64 24 0 48 0 80 20 96 0 32 1 48 0 104 0 112 0 120 0 128 1 0 136 1 0 152 1 0 72 2 82 84 10 0 16 0 24 0 74 16 8 0 16 0 26 0 32 0 40 0 48 0 56 0 64 0 90 0 122 4 8 0 32 0 128 1 0 136 1 0 144 1 0 152 1 0 162 1 6 8 0 18 0 24 0 168 1 0 178 1 0 186 1 0 192 1 0 200 1 0 208 1 0 224 1 0 233 1 0 0 0 0 0 0 0 0 96 0 106 33 10 11 10 5 97 100 109 105 110 16 2 24 2 10 10 10 4 114 111 111 116 16 2 24 2 18 4 114 111 111 116 24 2 128 1 1 136 1 3 152 1 0 184 1 0 194 1 45 83 69 76 69 67 84 32 120 32 70 82 79 77 32 100 97 116 97 98 97 115 101 95 49 48 56 55 53 49 46 112 117 98 108 105 99 46 116 95 49 48 56 55 53 49 200 1 156 1 232 1 0 242 1 4 8 0 18 0 248 1 0 128 2 0 146 2 0 154 2 0 178 2 0 184 2 1 192 2 157 1 200 2 0 224 2 0 128 3 0 136 3 1 168 3 0 176 3 0 208 3 0]
```
I'm worried that there's some bug around virtual table scans / joins that's just been uncovered.

--- a/pkg/ccl/logictestccl/testdata/logic_test/multi_region_foreign_key_lookup_join
+++ b/pkg/ccl/logictestccl/testdata/logic_test/multi_region_foreign_key_lookup_join
@@ -167,7 +167,7 @@ vectorized: true
     └── • scan
           missing stats
           table: child_rbr@child_rbr_pkey
-          spans: [/'ap-southeast-2' - /'ap-southeast-2'] [/'ca-central-1' - /'us-east-1']
+          spans: FULL SCAN
 
 query IIIIITI
 SELECT *
@@ -200,7 +200,7 @@ vectorized: true
         └── • scan
               missing stats
               table: child_rbr@child_rbr_pkey
-              spans: [/'ap-southeast-2' - /'ap-southeast-2'] [/'ca-central-1' - /'us-east-1']
+              spans: FULL SCAN
 
 query IIIIITI rowsort
 SELECT *

--- a/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_query_behavior
+++ b/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_query_behavior
@@ -1983,7 +1983,7 @@ SELECT * FROM [EXPLAIN INSERT INTO regional_by_row_table_virt (pk, a, b) VALUES 
 │                   └── • scan
 │                         missing stats
 │                         table: regional_by_row_table_virt@regional_by_row_table_virt_pkey
-│                         spans: [/'ap-southeast-2' - /'ap-southeast-2'] [/'ca-central-1' - /'us-east-1']
+│                         spans: FULL SCAN (SOFT LIMIT)
 │
 └── • constraint-check
     │
@@ -2162,7 +2162,7 @@ SELECT * FROM [EXPLAIN INSERT INTO regional_by_row_table_virt_partial (pk, a, b)
 │                   └── • scan
 │                         missing stats
 │                         table: regional_by_row_table_virt_partial@regional_by_row_table_virt_partial_pkey
-│                         spans: [/'ap-southeast-2' - /'ap-southeast-2'] [/'ca-central-1' - /'us-east-1']
+│                         spans: FULL SCAN (SOFT LIMIT)
 │
 └── • constraint-check
     │
@@ -2182,7 +2182,7 @@ SELECT * FROM [EXPLAIN INSERT INTO regional_by_row_table_virt_partial (pk, a, b)
                     └── • scan
                           missing stats
                           table: regional_by_row_table_virt_partial@regional_by_row_table_virt_partial_pkey
-                          spans: [/'ap-southeast-2' - /'ap-southeast-2'] [/'ca-central-1' - /'us-east-1']
+                          spans: FULL SCAN (SOFT LIMIT)
 
 query T retry
 SELECT * FROM [EXPLAIN UPSERT INTO regional_by_row_table_virt_partial (pk, a, b) VALUES (1, 1, 1)] OFFSET 2

--- a/pkg/sql/opt/xform/select_funcs.go
+++ b/pkg/sql/opt/xform/select_funcs.go
@@ -453,6 +453,31 @@ func (c *CustomFuncs) GenerateConstrainedScans(
 			return
 		}
 
+		// Make a best-effort check to avoid generating trivial constrained scans
+		// that actually scan the entire table.
+		//
+		// As a special case, omit singleton tables (statically guaranteed to have
+		// one row). This is advantageous because a full-table constrained scan for
+		// a singleton table will be performed using a Get instead of a Scan, which
+		// allows for some low-level optimizations.
+		if !grp.Relational().Cardinality.IsZeroOrOne() {
+			checkConstraintFilters := c.checkConstraintFilters(scanPrivate.Table)
+			for i := range checkConstraintFilters {
+				if !checkConstraintFilters[i].ScalarProps().TightConstraints {
+					continue
+				}
+				optionalConstraints := checkConstraintFilters[i].ScalarProps().Constraints
+				if optionalConstraints == nil {
+					continue
+				}
+				for j := 0; j < optionalConstraints.Length(); j++ {
+					if combinedConstraint.Contains(c.e.evalCtx, optionalConstraints.Constraint(j)) {
+						return
+					}
+				}
+			}
+		}
+
 		// Construct new constrained ScanPrivate.
 		newScanPrivate := *scanPrivate
 		newScanPrivate.Distribution.Regions = nil

--- a/pkg/sql/opt/xform/testdata/rules/groupby
+++ b/pkg/sql/opt/xform/testdata/rules/groupby
@@ -2704,23 +2704,23 @@ project
       │    │    ├── limit hint: 3.00
       │    │    ├── union-all
       │    │    │    ├── columns: a:2!null rowid:7!null
-      │    │    │    ├── left columns: a:48 rowid:53
-      │    │    │    ├── right columns: a:57 rowid:62
+      │    │    │    ├── left columns: a:30 rowid:35
+      │    │    │    ├── right columns: a:39 rowid:44
       │    │    │    ├── ordering: +2
       │    │    │    ├── limit hint: 3.00
       │    │    │    ├── scan regional@partial_a,partial
-      │    │    │    │    ├── columns: a:48!null rowid:53!null
-      │    │    │    │    ├── constraint: /47/48: [/'east' - /'east']
-      │    │    │    │    ├── key: (53)
-      │    │    │    │    ├── fd: (53)-->(48), (48)-->(53)
-      │    │    │    │    ├── ordering: +48
+      │    │    │    │    ├── columns: a:30!null rowid:35!null
+      │    │    │    │    ├── constraint: /29/30: [/'east' - /'east']
+      │    │    │    │    ├── key: (35)
+      │    │    │    │    ├── fd: (35)-->(30), (30)-->(35)
+      │    │    │    │    ├── ordering: +30
       │    │    │    │    └── limit hint: 3.00
       │    │    │    └── scan regional@partial_a,partial
-      │    │    │         ├── columns: a:57!null rowid:62!null
-      │    │    │         ├── constraint: /56/57: [/'west' - /'west']
-      │    │    │         ├── key: (62)
-      │    │    │         ├── fd: (62)-->(57), (57)-->(62)
-      │    │    │         ├── ordering: +57
+      │    │    │         ├── columns: a:39!null rowid:44!null
+      │    │    │         ├── constraint: /38/39: [/'west' - /'west']
+      │    │    │         ├── key: (44)
+      │    │    │         ├── fd: (44)-->(39), (39)-->(44)
+      │    │    │         ├── ordering: +39
       │    │    │         └── limit hint: 3.00
       │    │    └── aggregations
       │    │         └── count-rows [as=count_rows:10]

--- a/pkg/sql/opt/xform/testdata/rules/insert
+++ b/pkg/sql/opt/xform/testdata/rules/insert
@@ -175,17 +175,15 @@ insert t
       │         ├── columns: t.k:26!null t.r:27!null t.a:28!null t.b:29 t.c:30
       │         ├── key: (26)
       │         ├── fd: ()-->(28), (26)-->(27,29,30)
-      │         ├── index-join t
+      │         ├── scan t
       │         │    ├── columns: t.k:26!null t.r:27!null t.a:28 t.b:29 t.c:30
+      │         │    ├── check constraint expressions
+      │         │    │    └── t.r:27 IN ('east', 'west') [outer=(27), constraints=(/27: [/'east' - /'east'] [/'west' - /'west']; tight)]
+      │         │    ├── computed column expressions
+      │         │    │    └── t.b:29
+      │         │    │         └── t.k:26 % 9
       │         │    ├── key: (26)
-      │         │    ├── fd: (26)-->(27-30)
-      │         │    └── scan t@t_r_c_idx
-      │         │         ├── columns: t.k:26!null t.r:27!null t.c:30
-      │         │         ├── constraint: /27/30/26
-      │         │         │    ├── [/'east' - /'east']
-      │         │         │    └── [/'west' - /'west']
-      │         │         ├── key: (26)
-      │         │         └── fd: (26)-->(27,30)
+      │         │    └── fd: (26)-->(27-30)
       │         └── filters
       │              └── t.a:28 = 10 [outer=(28), constraints=(/28: [/10 - /10]; tight), fd=()-->(28)]
       └── fast-path-unique-checks-item: t(c)
@@ -525,19 +523,15 @@ insert t
       │         ├── key: (22)
       │         └── fd: ()-->(24), (22)-->(23,25)
       ├── fast-path-unique-checks-item: t(b)
-      │    └── select
+      │    └── index-join t
       │         ├── columns: t.k:38!null t.r:39!null t.a:40 t.b:41!null
       │         ├── key: (38)
       │         ├── fd: ()-->(41), (38)-->(39,40)
-      │         ├── scan t@t_r_a_b_idx
-      │         │    ├── columns: t.k:38!null t.r:39!null t.a:40 t.b:41
-      │         │    ├── constraint: /39/40/41/38
-      │         │    │    ├── [/'east' - /'east']
-      │         │    │    └── [/'west' - /'west']
-      │         │    ├── key: (38)
-      │         │    └── fd: (38)-->(39-41)
-      │         └── filters
-      │              └── t.b:41 = 2 [outer=(41), constraints=(/41: [/2 - /2]; tight), fd=()-->(41)]
+      │         └── scan t@t_b_a_idx
+      │              ├── columns: t.k:38!null t.a:40 t.b:41!null
+      │              ├── constraint: /41/40/38: [/2 - /2]
+      │              ├── key: (38)
+      │              └── fd: ()-->(41), (38)-->(40)
       ├── fast-path-unique-checks-item: t(r,a,b)
       │    └── scan t@t_r_a_b_idx
       │         ├── columns: t.k:54!null t.r:55!null t.a:56!null t.b:57!null

--- a/pkg/sql/opt/xform/testdata/rules/join
+++ b/pkg/sql/opt/xform/testdata/rules/join
@@ -13050,7 +13050,6 @@ explain
            │    │    │    ├── columns: u85353.a:6 u85353.b:7
            │    │    │    └── scan u85353@b_idx
            │    │    │         ├── columns: u85353.b:7 u85353.rowid:10!null
-           │    │    │         ├── constraint: /9/7/10: [/0 - /7]
            │    │    │         ├── flags: force-index=b_idx
            │    │    │         ├── key: (10)
            │    │    │         └── fd: (10)-->(7)

--- a/pkg/sql/opt/xform/testdata/rules/select
+++ b/pkg/sql/opt/xform/testdata/rules/select
@@ -133,6 +133,15 @@ CREATE TABLE virtual (
 )
 ----
 
+exec-ddl
+CREATE TABLE singleton_check(
+  singleton BOOL DEFAULT TRUE,
+  count INT NOT NULL,
+  PRIMARY KEY (singleton),
+  CHECK (singleton)
+)
+----
+
 # --------------------------------------------------
 # GeneratePartialIndexScans
 # --------------------------------------------------
@@ -2609,7 +2618,8 @@ project
       ├── fd: ()-->(3)
       ├── scan t114250
       │    ├── columns: x:1!null z:3!null
-      │    └── constraint: /1/2: [/1 - /4]
+      │    └── check constraint expressions
+      │         └── (x:1 > 0) AND (x:1 < 5) [outer=(1), constraints=(/1: [/1 - /4]; tight)]
       └── filters
            └── z:3 = 5 [outer=(3), constraints=(/3: [/5 - /5]; tight), fd=()-->(3)]
 
@@ -2649,6 +2659,21 @@ project
       ├── constraint: /4/1: [/true - /true]
       └── key: (1)
 
+# Regression test for #114470 - allow a non-selective constraint for singleton
+# tables.
+#
+# This is a simplified version of an internal query performed on the
+# system.span_count table, which is vulnerable to the small increase in latency
+# caused by performing a KV Scan instead of a Get.
+opt expect=GenerateConstrainedScans
+SELECT * FROM singleton_check WHERE singleton;
+----
+scan singleton_check
+ ├── columns: singleton:1!null count:2!null
+ ├── constraint: /1: [/true - /true]
+ ├── cardinality: [0 - 1]
+ ├── key: ()
+ └── fd: ()-->(1,2)
 
 # --------------------------------------------------
 # GenerateInvertedIndexScans

